### ISSUE TITLE
UK postcode polygons from wikipedia

### ISSUE THREAD

**Issue by missinglink** [whosonfirst-data/whosonfirst-data#540]
heya, there are a bunch of Uk postcode polygons in wikipedia in KML format.

I downloaded them and converted them to geojson and stuck them here:
https://github.com/missinglink/uk-postcode-polygons

what are the general feelings about importing this sort of thing in to WOF?

**Comment by stepps00** [comment]
Hey @missinglink, thanks for filing the issue.

As the wikipedia page suggests, it looks like this data is district-level groupings of `postalcodes`, rather than the individual `postalcode` records that Who's On First currently maintains. Not to say this data can't be imported, but it would likely fit into a placetype that Who's On First does not currently have. 

For example:

[This geojson file](https://github.com/missinglink/uk-postcode-polygons/blob/master/geojson/BN.geojson) contains records like `BN1`. Who's On First, though, has [these](https://whosonfirst.mapzen.com/spelunker/search/?q=BN1#3/28.54/-0.09) records that all fall within `BN1`.

I will poke around, but have you come across more granular `postalcode` data in Wikipedia? We're happy to update `postalcode` data, as we've recently done in [France](https://github.com/whosonfirst-data/whosonfirst-data-postalcode-fr) and [Australia](https://github.com/whosonfirst-data/whosonfirst-data-postalcode-au).

**Comment by nvkelso** [comment]
Would the 2 char prefix postcode be useful? Of course any shape is useful! But would, say, Pelias use it to display as the postcode, even when it's not fully formed? Seems like it _may_ be useful for data visualization (in the USA it's usually the first 3 char prefix), but I'm leery to add more complexity to the system right now.

Screenshots of what @stepps00 describes:

**Who's On First** has records like `BN1 3HH`

<img width="1186" alt="screen shot 2016-12-01 at 15 57 48" src="https://cloud.githubusercontent.com/assets/853051/20817906/59d04be2-b7df-11e6-948e-8f6a9e55106f.png">

**Wikipedia** has areas, but they're aggregated up substantially to only the initial 2 prefix `BN`.

<img width="548" alt="screen shot 2016-12-01 at 15 58 58" src="https://cloud.githubusercontent.com/assets/853051/20817907/59d07f22-b7df-11e6-963b-932a48146223.png">

**Comment by missinglink** [comment]
This request originated from a community contributor who was trying to perform postcode validation using Pelias for addresses in the UK.

Being able to validate that the first 3-4 characters are in the correct geography is generally useful in the UK where polygon and address data is not so easy to come by.

The Pelias data model didn't consider the idea of 'partial postcodes' and I guess WOF and other systems rarely support that sort of thing, I'm not sure how useful it is to people in general?

I've personally wanted open postcode polygons for the UK for years now and AFAIK they don't exist in the open?

I will have some free cycles over the holiday break, I considered using address point data to create convex hulls of the postal boundaries for the UK, it wouldn't be perfect but it would certainly be better than centroid-only data.

Would you consider merging UK postcode polygons computed using a convex hull algorithm where the points were sourced from OSM and other sources?

I'm not sure of the licencing around that, as it is likely considered a "derivative work" under the ODBL, it might need to be published as a separate data set.

**Comment by missinglink** [comment]
also @nvkelso the wikipedia ones go down to 3-4 characters as per https://github.com/missinglink/uk-postcode-polygons/blob/master/geojson/BN.geojson (you can click on individual area in there to see more depth)

**Comment by nvkelso** [comment]
There's a compelling argument that 3-4 postcode chars for commonwealth postcodes are equivalent to ZIP, where full 6 char commonwealth is more equivalent to ZIP + 4.

But the Share Alike terms in Wikipedia make me leery about including this directly into Who's On First's CC-BY terms.

**Comment by nvkelso** [comment]
We might get good results looking at the venues data in Who's On First and using `clustr` (versus getting caught up in ODbL of OpenStreetMap data). @thisisaaronland knows more.

**Comment by thisisaaronland** [comment]
I agree with @nvkelso that the "share alike" license introduces restrictions that WOF actively avoids.

The Clustr stuff could work well enough to try, in a weird data is better than no data kind of way. Basically we just need to plough through all the GB postal codes and bucket their lat/lon pairs by n-leading characters in the actual postal code and then feed those files to Clustr.

**Comment by nvkelso** [comment]
Related: https://github.com/openaddresses/openaddresses/issues/2299 for UK Data - OpenAddressesUK.Org. 

For example, https://alpha.openaddressesuk.org/addresses/fVEYss.json has (only) a lat/lon for the postcode. The are inputs listed, as well.
